### ISSUE TITLE
CLI '-s' option does not match by data from '-t'

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#7523]
https://checkstyle.org/cmdline.html#Command_line_usage

```
$ cat Test.java 
public class Test {
  public Test() {}
}

$ java -jar /var/tmp/checkstyle-8.29-all.jar -t Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--CTOR_DEF -> CTOR_DEF [2:2]
    |   |--MODIFIERS -> MODIFIERS [2:2]
    |   |   `--LITERAL_PUBLIC -> public [2:2]
    |   |--IDENT -> Test [2:9]
    |   |--LPAREN -> ( [2:13]
    |   |--PARAMETERS -> PARAMETERS [2:14]
    |   |--RPAREN -> ) [2:14]
    |   `--SLIST -> { [2:16]
    |       `--RCURLY -> } [2:17]
    `--RCURLY -> } [3:0]

$ java -jar /var/tmp/checkstyle-8.29-all.jar -s 2:17 Test.java 
/CLASS_DEF[./IDENT[@text='Test']]/OBJBLOCK/CTOR_DEF[./IDENT[@text='Test']]/SLIST

$ java -jar /var/tmp/checkstyle-8.29-all.jar -s 2:16 Test.java 

$ 
```

by first command, xpath is printed for parent token `SLIST`, but expected `RCURLY`.
by second command, nothing is printed but expected xpath for `SLIST`.

**Comment by romani** [comment]
@timurt , can you confirm that this valid issue ?

**Comment by gaurabdg** [comment]
I am on it.

**Comment by gaurabdg** [comment]
@romani I checked the line, column pair in vscode and it's consistent with the "-s" parameter ouput but inconsistent  with "-t". I guess it's just about an offset of 1 in `AstTreeStringPrinter`. Please correct me if I am wrong as it will change a lot of UTs.

**Comment by romani** [comment]
yes, we need  to make it consistent, and what we print in CLI should be reusable as is in `-s`.

**Comment by gaurabdg** [comment]
Why is the column number 0 based in implementations but 1 based in xpathquery generator?

**Comment by gaurabdg** [comment]
@romani @timurt Do you want me to change the AST implementations' column number to be 1 based or adapt `-s` and make it zero based in the background?

**Comment by romani** [comment]
Looks like related https://github.com/checkstyle/checkstyle/issues/4997

@timurt , please provide history context of this problem

**Comment by gaurabdg** [comment]
I am following the deliverables mentioned in #4997's comment.

**Comment by gaurabdg** [comment]
Should I resolve as mentioned in #4997 or just make an offset adjustment for cli parameters since there are almost 400 UT errors and will take a lot of time

**Comment by timurt** [comment]
@romani I tried to recall details, but I think @MEZk 's description in [#4997](https://github.com/checkstyle/checkstyle/issues/4997#issue-252697857) contains all required information regarding this issue

`AstTreeStringPrinter` prints columns starting from 0 and does not consider tabulations
`XpathQueryGenerator` is 1-based and calculates real column number considering tabulation
`AbstractCheck` is 1-based and calculates real column number considering tabulation

Simply, `-s` uses same column format as Checkstyle violations, because we wanted user to have opportunity to generate xpath queries from violations he faced.

`-t` prints AST tree in 0-based column format. 

To make decision we need to understand why `AbstractCheck` and `AstTreeStringPrinter` do not print columns in same format. May be there were some reasons behind this solution. 

If we decide, that we want to make all column numbers consistent (AST tree, xpath queries, violations), as easiest solution I would recommend to change `AstTreeStringPrinter` with same real column number calculations as we used in `AbstractCheck` (`final int col = 1 + CommonUtils.lengthExpandedTabs(
            getLines()[lineNo - 1], colNo, tabWidth);`)

@rnveach please share your thoughts

**Comment by rnveach** [comment]
> To make decision we need to understand why AbstractCheck and AstTreeStringPrinter do not print columns in same format

This was probably so column number reports column position that you would see in an IDE, which is also 1 based.

> If we decide, that we want to make all column numbers consistent 

I feel we should make them consistent, regardless of 0 or 1 based.

I am leaning towards 1 based since that is what will make the most sense to users. All IDEs and notepads use 1 based positioning and they would rather see number and find the exact same number in their editor. Users just want the simplest path without much thought.

**Comment by romani** [comment]
yes, `1` based is what we need, as checkstyle suppose to be used in IDEs a lot. Natural positioning is good.

**Comment by rnveach** [comment]
@romani I assume main fix should be in https://github.com/checkstyle/checkstyle/issues/4997 and this issue will just be to ensure the problem here is fixed?

**Comment by romani** [comment]
@gaurabdg , I changed label from easy to hard.
With current amount of active PRs it will be hard to update index from 0 to 1 in DetailAst, too much conflict will be.

We will come back to this issue when we start main phase of GSoC, resolving of this issue will be required for xpath related proejcts.

**Comment by gaurabdg** [comment]
Sure.

**Comment by ayushactiveat** [comment]
hi @romani,
i am interested in working on this issue, also i will be submitting a proposal for the XPath-related project for GSoC 2026 so i think resolving this cli inconsistency would be a great way to prepare and prove myself for the project.
Please let me know if I can take this up.

**Comment by romani** [comment]
@ayushactiveat , I highly recommend to finish one PR for each level "good xxxx issue" .
This issue will wait for you, if you already finished good xxxxx issues, please start this.

**Comment by romani** [comment]
Fix is merged.
@ayushactiveat thanks a lot for yours hary work
